README.md
# D3 Skeletons

This extension contains javascript code snippets for D3.js on Visual Studio Code. It is meant to provide high level skeletons of common D3 Templates. It also attempts to provide comments that helps understand the templated code to make adjustments as necessary.

# Snippets
**⇥** means the `TAB` key

| Trigger  | Content |
| -------: | ------- |
| `d3s.lineChart→`   | https://bl.ocks.org/mbostock/3883245 |
| `d3s.pieChart→`  | https://bl.ocks.org/mbostock/3887235 |
| `d3s.barChart→`  | https://bl.ocks.org/mbostock/3885304|

Created by Noah Ezekwugo & Lee Polla

CHANGELOG.md
# Change Log
All notable changes to the "d3-skeletons" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [0.0.1]
- Initial release
- Added Line Chart
- Added Bar Chart
- Added Pie Chart

snippets/barChart.js
var svg = d3.select("svg"),
    margin = {top: 20, right: 20, bottom: 30, left: 40},
    width = +svg.attr("width") - margin.left - margin.right,
    height = +svg.attr("height") - margin.top - margin.bottom;

//Generates a categorical x axis and a numeric y axis
var x = d3.scaleBand().rangeRound([0, width]).padding(0.1),
    y = d3.scaleLinear().rangeRound([height, 0]);

//Generates chart container that is adjusted by margin parameters
var g = svg.append("g")
    .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

//Reads data from tsv
d3.tsv("data.tsv", function(d) {
  //set up your data transforms here
  d.name = "";
  d.value = + 0;
  return d;
}, function(error, data) {
  if (error) throw error;

  //Sets categorical x domain based on given column
  x.domain(data.map(function(d) { return d.name; }));
  y.domain([0, d3.max(data, function(d) { return d.value; })]);

  //Draw x axis and adjust for container size
  g.append("g")
      .attr("class", "axis axis--x")
      .attr("transform", "translate(0," + height + ")")
      .call(d3.axisBottom(x));
  //Draw y axis with various styling and transforms
  g.append("g")
      .attr("class", "axis axis--y")
      .call(d3.axisLeft(y).ticks(10, "%"))
    .append("text")
      .attr("transform", "rotate(-90)")
      .attr("y", 6)
      .attr("dy", "0.71em")
      .attr("text-anchor", "end")
      .text("Value");
  //Draw bars for each name and value in data
  g.selectAll(".bar")
    .data(data)
    .enter().append("rect")
      .attr("class", "bar")
      .attr("x", function(d) { return x(d.name); })
      .attr("y", function(d) { return y(d.value); })
      .attr("width", x.bandwidth())
      .attr("height", function(d) { return height - y(d.value); });//draws bars on screen from top to bottom
});

snippets/lineChart.js
var svg = d3.select('svg'),
margin = {top: 20, right: 20, bottom: 30, left: 50},
width = +svg.attr('width') - margin.left - margin.right,
height = +svg.attr('height') - margin.top - margin.bottom,
g = svg.append('g').attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');

//Creates object for formatting date strings
var parseTime = d3.timeParse('%d-%b-%y');

//Prepares to draw x values from left to right
var x = d3.scaleTime()
  .rangeRound([0, width]);

//Prepares to draw y values from top to bottom
var y = d3.scaleLinear()
  .rangeRound([height, 0]);

//Declare your numeric x and y variables here
var line = d3.line()
  .x(function(d) { return x(d.date); })
  .y(function(d) { return y(d.close); });

//Data reading section, change to csv if necessary
d3.tsv('data.tsv', function(d) {
  d.date = parseTime(d.date);
  d.close = +d.close;
  return d;
}, function(error, data) {
  if (error) throw error;

//Sets min and max domain values based from the given data
x.domain(d3.extent(data, function(d) { return d.date; }));
y.domain(d3.extent(data, function(d) { return d.close; }));

  //Draw x axis with adjustments for positioning
  g.append('g')
      .attr('transform', 'translate(0,' + height + ')')
      .call(d3.axisBottom(x))
    .select('.domain')
      .remove();
  //Draw y axis with various styling
  g.append('g')
      .call(d3.axisLeft(y))
    .append('text')
      .attr('fill', '#000')
      .attr('transform', 'rotate(-90)')
      .attr('y', 6)
      .attr('dy', '0.71em')
      .attr('text-anchor', 'end')
      .text('Price ($)');

  //Add lines to the drawing area
  g.append('path')
    .datum(data)
    .attr('fill', 'none')
    .attr('stroke', 'steelblue')
    .attr('stroke-linejoin', 'round')
    .attr('stroke-linecap', 'round')
    .attr('stroke-width', 1.5)
    .attr('d', line); //Defines lines based on the x and y values defined earlier
});

snippets/pieChart.js
// Template taken from https://bl.ocks.org/mbostock/3887235
//dimensions declarations
var width = 650;
var height = 400;
var margin = {top: 20, right: 15, bottom: 30, left: 40};
var w = width - margin.left - margin.right;
var h = height - margin.top - margin.bottom;

//dataset declarations
var csv = "flavors_of_cacao.csv";//Your CSV file here;
var dataset;

d3.csv(csv, function(error, data) {
    //read in the data
    if (error) return console.warn(error);
        data.forEach(function(d) {
            d.name = d.Company; //Your data name here
            d.value1 = + d.Rating; //Your data value here
            d.value2 = + d.CocoaPercent.slice(0, 2)
    });
    dataset = data;    

//pie charts

var svg = d3.select('body').append('svg'),
    width = +svg.attr("width"),
    height = +svg.attr("height"),
    radius = Math.min(width, height) / 2,
    g = svg.append("g").attr("transform", "translate(" + width / 2 + "," + height / 2 + ")");

var color = d3.scaleOrdinal(["#98abc5", "#8a89a6", "#7b6888", "#6b486b", "#a05d56", "#d0743c", "#ff8c00"]);

var pie = d3.pie()
    .sort(null)
    .value(function(d) { return d.value2; });

//Calculate pie chart arc  geometries
var path = d3.arc()
    .outerRadius(radius - 10)
    .innerRadius(0);

//radius offsets put label to the edge of pie
var label = d3.arc()
    .outerRadius(radius - 40)
    .innerRadius(radius - 40);

//Generate pie slices based on data
var arc = g.selectAll(".arc")
    .data(pie(dataset))
    .enter().append("g")
    .attr("class", "arc");

    //Color the pie segments based on unique names
    arc.append("path")
      .attr("d", path)
      .attr("fill", function(d) { return color(d.name); });

    //Adds text to pie chart
    arc.append("text")
      .attr("transform", function(d) { return "translate(" + label.centroid(d) + ")"; })
      .attr("dy", "0.35em")
      .text(function(d) { return d.value2; });
});